### DISCUSSION TITLE
no-prototype-builtins & prefer-object-has-own

### DISCUSSION THREAD

**Discussion by erosman** [eslint/eslint#17751] [Q&A]
It is not an issue but I have noticed that with "no-prototype-builtins" enabled (i.e. with "extends": "eslint:recommended"), "prefer-object-has-own" is never shown (one is a subset of the other).
Is that expected?
Would that make "prefer-object-has-own" redundant (with "eslint:recommended")?

**Comment by Rec0iL99** [comment]
I don't think so. `no-prototype-builtins` is a rule that enforces using methods from `Object.prototype` instead of calling these methods directly on objects.

```js
/* eslint no-prototype-builtins: "error" */
/* eslint prefer-object-has-own: "error" */

let hasBarProperty = Object.hasOwn(foo, "bar");
```

Here `no-prototype-builtins` is enforcing doing `Object.prototype.hasOwnProperty.call(foo, "bar")` instead of `foo.hasOwnProperty("bar")`, while the `prefer-object-has-own` enforces usage of  `Object.hasOwn(foo, "bar")` instead of `Object.prototype.hasOwnProperty.call(foo, "bar")`.

**Reply by erosman** [reply]
I should have included a code example. 
```js
/*eslint no-prototype-builtins: "error"*/
/*eslint prefer-object-has-own: "error"*/

const obj = {a: 1, b: 2};

console.log(obj.hasOwnProperty('a'));
```

> Do not access Object.prototype method 'hasOwnProperty' from target object.  ([no-prototype-builtins](https://eslint.org/docs/rules/no-prototype-builtins))

**Reply by Rec0iL99** [reply]
Could you please provide more context?

**Reply by erosman** [reply]
As mentioned, this is not a problem. It is merely a curiosity. To start with, "prefer-object-has-own" does not generate a message in the following.

```js
/*eslint prefer-object-has-own: "error"*/

const obj = {a: 1, b: 2};

console.log(obj.hasOwnProperty('a'));
```

**Reply by Rec0iL99** [reply]
`prefer-object-has-own` rule only enforces using `Object.hasOwn(obj, 'a')` instead of `Object.prototype.hasOwnProperty.call(obj, 'a')` since `Object.hasOwn` is an alias for  `Object.prototype.hasOwnProperty.call`.

**Reply by erosman** [reply]
However ....

> [Object.hasOwn()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn)
>
> Note: `Object.hasOwn()` is intended as a replacement for [Object.prototype.hasOwnProperty()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty).

IMHO, the rule should also mark `obj.hasOwnProperty('a')`.

```js
/*eslint no-prototype-builtins: "error"*/
/*eslint prefer-object-has-own: "error"*/

const obj = {a: 1, b: 2};

obj.hasOwnProperty('a');
Object.hasOwnProperty.call(obj, 'a');
Object.prototype.hasOwnProperty(obj, 'a');
Object.prototype.hasOwnProperty.call(obj, 'a');
```
